# Live Code 1

- Nama  : Pramudya Adi Nugraha
- Batch : 10

## Description

### Objective
- Mampu memuat data dengan Pandas
- Mampu menerapkan looping for dalam suatu kasus
- Mampu menerapkan konsep conditional if dalam suatu kasus
- Mampu menerapkan function pada sebuah kasus
- Mampu mengimplementasikan pandas query dalam pengolahan data

### Problem Statement
Kamu adalah seorang data scientist yang bekerja di salah satu perusahaan manufaktur baterai Lithium ion. Tim produksi meminta kamu untuk mengecek apakah baterai yang mereka buat dalam kondisi bagus atau tidak. Sebelum melakukan pengecekan lebih lanjut, kamu diminta untuk menghitung hambatan/resistansi dan Daya setiap detik:
R=\frac{V}{I}

P=VI

R adalah Kapasitas dalam satuan Ohm
P adalah Daya dalam Watt
I adalah kuat arus dalam satuan Ampere
V adalah tegangan dalam satuan Volt

## Working Area

a. Buatlah dua fungsi eksplisit masing-masing memuat perhitungan hambatan (R) dan daya (P) yang nantinya akan digunakan untuk perhitungan

Import data

In [2]:
import pandas as pd

Data Loading

In [63]:
data = pd.read_excel('NMC1-9.xlsx')
data

time(s)    V_m  Current (A)  Cycle       Cap
0            0  4.098        0.997      1  2.840817
1            1  4.096        0.998      1  2.840817
2            2  4.094        0.999      1  2.840817
3            3  4.093        0.998      1  2.840817
4            4  4.092        0.999      1  2.840817
...        ...    ...          ...    ...       ...
90977    90977  2.506        1.002      9  2.770180
90978    90978  2.504        1.002      9  2.770180
90979    90979  2.502        1.002      9  2.770180
90980    90980  2.500        1.002      9  2.770180
90981    90981  2.497        1.002      9  2.770180

[90982 rows x 5 columns]

In [5]:
data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 90982 entries, 0 to 90981
Data columns (total 5 columns):
 #   Column       Non-Null Count  Dtype  
---  ------       --------------  -----  
 0   time(s)      90982 non-null  int64  
 1   V_m          90982 non-null  float64
 2   Current (A)  90982 non-null  float64
 3   Cycle        90982 non-null  int64  
 4   Cap          90982 non-null  float64
dtypes: float64(3), int64(2)
memory usage: 3.5 MB


Answer

In [7]:
def hambatan(x, y):
    r = x/y
    return r

In [9]:
def daya(a, b):
    p = a/b
    return p

Hasil

In [13]:
hambatan(data['V_m'],data['Current (A)'])

0        4.110331
1        4.104208
2        4.098098
3        4.101202
4        4.096096
           ...   
90977    2.500998
90978    2.499002
90979    2.497006
90980    2.495010
90981    2.492016
Length: 90982, dtype: float64

In [14]:
daya(data['V_m'],data['Current (A)'])

0        4.110331
1        4.104208
2        4.098098
3        4.101202
4        4.096096
           ...   
90977    2.500998
90978    2.499002
90979    2.497006
90980    2.495010
90981    2.492016
Length: 90982, dtype: float64

b. Buat kolom baru dengan nama kolom R/P berisikan hasil perhitungan daya/hambatan dengan kriteria untuk Kapasitas >=2.5 hitung daya, Kapasitas <2.5 hitung nilai hambatan. (Hint: Gunakan for loop dan if statement untuk mengerjakan perintah nomor b, kamu bisa menggunakan list untuk menyimpan perhitungan sementara, setelahnya dapat dimasukkan ke kolom R/P).

Import data

In [25]:
import pandas as pd

Data Loading

In [62]:
data = pd.read_excel('NMC1-9.xlsx')
data

time(s)    V_m  Current (A)  Cycle       Cap
0            0  4.098        0.997      1  2.840817
1            1  4.096        0.998      1  2.840817
2            2  4.094        0.999      1  2.840817
3            3  4.093        0.998      1  2.840817
4            4  4.092        0.999      1  2.840817
...        ...    ...          ...    ...       ...
90977    90977  2.506        1.002      9  2.770180
90978    90978  2.504        1.002      9  2.770180
90979    90979  2.502        1.002      9  2.770180
90980    90980  2.500        1.002      9  2.770180
90981    90981  2.497        1.002      9  2.770180

[90982 rows x 5 columns]

In [24]:
data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 90982 entries, 0 to 90981
Data columns (total 5 columns):
 #   Column       Non-Null Count  Dtype  
---  ------       --------------  -----  
 0   time(s)      90982 non-null  int64  
 1   V_m          90982 non-null  float64
 2   Current (A)  90982 non-null  float64
 3   Cycle        90982 non-null  int64  
 4   Cap          90982 non-null  float64
dtypes: float64(3), int64(2)
memory usage: 3.5 MB


Answer

In [29]:
r_p = []

for i in range(len(data)):
    if data['Cap'][i] >= 2.5:
        r_p.append(daya(data['V_m'][i],data['Current (A)'][i]))
    elif data['Cap'][i] < 2.5:
        r_p.append(hambatan(data['V_m'][i],data['Current (A)'][i]))

Hasil

In [30]:
data['R/P'] = r_p
data

time(s)    V_m  Current (A)  Cycle       Cap       R/P
0            0  4.098        0.997      1  2.840817  4.110331
1            1  4.096        0.998      1  2.840817  4.104208
2            2  4.094        0.999      1  2.840817  4.098098
3            3  4.093        0.998      1  2.840817  4.101202
4            4  4.092        0.999      1  2.840817  4.096096
...        ...    ...          ...    ...       ...       ...
90977    90977  2.506        1.002      9  2.770180  2.500998
90978    90978  2.504        1.002      9  2.770180  2.499002
90979    90979  2.502        1.002      9  2.770180  2.497006
90980    90980  2.500        1.002      9  2.770180  2.495010
90981    90981  2.497        1.002      9  2.770180  2.492016

[90982 rows x 6 columns]

c. Hitunglah rata-rata nilai kapasitas untuk arus = 1 Ampere dan cycle ganjil.

Import

In [35]:
import pandas as pd

Data Loading

In [61]:
data2 = pd.read_excel('NMC1-9.xlsx')
data2

time(s)    V_m  Current (A)  Cycle       Cap
0            0  4.098        0.997      1  2.840817
1            1  4.096        0.998      1  2.840817
2            2  4.094        0.999      1  2.840817
3            3  4.093        0.998      1  2.840817
4            4  4.092        0.999      1  2.840817
...        ...    ...          ...    ...       ...
90977    90977  2.506        1.002      9  2.770180
90978    90978  2.504        1.002      9  2.770180
90979    90979  2.502        1.002      9  2.770180
90980    90980  2.500        1.002      9  2.770180
90981    90981  2.497        1.002      9  2.770180

[90982 rows x 5 columns]

In [37]:
data2.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 90982 entries, 0 to 90981
Data columns (total 5 columns):
 #   Column       Non-Null Count  Dtype  
---  ------       --------------  -----  
 0   time(s)      90982 non-null  int64  
 1   V_m          90982 non-null  float64
 2   Current (A)  90982 non-null  float64
 3   Cycle        90982 non-null  int64  
 4   Cap          90982 non-null  float64
dtypes: float64(3), int64(2)
memory usage: 3.5 MB


Answer

In [66]:
data2[data2['Current (A)'] == 1]['Cap']

15       2.840817
19       2.840817
20       2.840817
21       2.840817
31       2.840817
           ...   
90909    2.770180
90923    2.770180
90937    2.770180
90949    2.770180
90967    2.770180
Name: Cap, Length: 11986, dtype: float64

In [67]:
data2[data2['Cycle']%2 != 0]['Cap']

0        2.840817
1        2.840817
2        2.840817
3        2.840817
4        2.840817
           ...   
90977    2.770180
90978    2.770180
90979    2.770180
90980    2.770180
90981    2.770180
Name: Cap, Length: 50540, dtype: float64

In [56]:
data2.loc[(data2['Current (A)'] == 1) & data2['Cycle']%2 != 0, 'Cap' ]

15       2.840817
19       2.840817
20       2.840817
21       2.840817
31       2.840817
           ...   
90909    2.770180
90923    2.770180
90937    2.770180
90949    2.770180
90967    2.770180
Name: Cap, Length: 6830, dtype: float64

In [54]:
rata2_cap = data2.loc[(data2['Current (A)'] == 1) & data2['Cycle']%2 != 0, 'Cap' ].mean()

Hasil

In [55]:
rata2_cap

2.8064845553925855

## Conclusions, Assumptions, Overall Analysis

- Dari perhitungan daya didapatkan hasil sebagai berikut : 

In [57]:
daya(data['V_m'],data['Current (A)'])

0        4.110331
1        4.104208
2        4.098098
3        4.101202
4        4.096096
           ...   
90977    2.500998
90978    2.499002
90979    2.497006
90980    2.495010
90981    2.492016
Length: 90982, dtype: float64

- Dari perhitungan hambatan didapatkan hasil sebagai berikut:

In [58]:
hambatan(data['V_m'],data['Current (A)'])

0        4.110331
1        4.104208
2        4.098098
3        4.101202
4        4.096096
           ...   
90977    2.500998
90978    2.499002
90979    2.497006
90980    2.495010
90981    2.492016
Length: 90982, dtype: float64

- Hasil perhitungan daya/hambatan dengan kriteria untuk Kapasitas >=2.5 hitung daya, Kapasitas <2.5 hitung nilai hambatan sebagai berikut:

In [60]:
data['R/P'] = r_p
data

time(s)    V_m  Current (A)  Cycle       Cap       R/P
0            0  4.098        0.997      1  2.840817  4.110331
1            1  4.096        0.998      1  2.840817  4.104208
2            2  4.094        0.999      1  2.840817  4.098098
3            3  4.093        0.998      1  2.840817  4.101202
4            4  4.092        0.999      1  2.840817  4.096096
...        ...    ...          ...    ...       ...       ...
90977    90977  2.506        1.002      9  2.770180  2.500998
90978    90978  2.504        1.002      9  2.770180  2.499002
90979    90979  2.502        1.002      9  2.770180  2.497006
90980    90980  2.500        1.002      9  2.770180  2.495010
90981    90981  2.497        1.002      9  2.770180  2.492016

[90982 rows x 6 columns]

- Untuk arus dengan besar = 1 Ampere dan Cycle ganjil, diperoleh rata-rata kapasitas sebesar 2.8064845553925855
- Terdapat 6830 data kapasitas yang memiliki arus sebesar 1 ampere dan termasuk dalam cycle ganjil